### PR TITLE
feat: optimism-specific receipt

### LINKED ISSUE #12588: Define `OpReceipt`
### Describe the feature

Define `OpReceipt` which should be a wrapper around the `op-alloy` type, and should impl `reth_primitives_traits::Receipt`.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by klkvr**
Closes https://github.com/paradigmxyz/reth/issues/12588

Same as #13295 but for OP receipt

### CODE REVIEW

File: crates/optimism/primitives/src/receipt.rs [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+/// Typed ethereum transaction receipt.
+/// Receipt containing result of transaction execution.
+#[derive(Clone, Debug, PartialEq, Eq, Default)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "reth-codec", derive(reth_codecs::CompactZstd))]
+#[cfg_attr(feature = "reth-codec", reth_codecs::add_arbitrary_tests)]
+#[cfg_attr(feature = "reth-codec", reth_zstd(
+    compressor = reth_zstd_compressors::RECEIPT_COMPRESSOR,
+    decompressor = reth_zstd_compressors::RECEIPT_DECOMPRESSOR
+))]
```
Comment (emhane) [inline crates/optimism/primitives/src/receipt.rs:20]:
```suggestion
#[cfg_attr(feature = "reth-codec", derive(reth_codecs::CompactZstd), reth_codecs::add_arbitrary_tests, reth_zstd(
    compressor = reth_zstd_compressors::RECEIPT_COMPRESSOR,
    decompressor = reth_zstd_compressors::RECEIPT_DECOMPRESSOR
))]
```

**Review by emhane — COMMENTED** [review]
is it not possible to make an enum with one variant wrapping `op_alloy_consensus::OpDepositReceipt`?

**Comment by klkvr** [comment]
this would require a manual compact impl through a struct same as the current one for compatibility

**Comment by klkvr** [comment]
@emhane I've created a draft using an enum https://github.com/paradigmxyz/reth/pull/13334, it needs an op-alloy release

I think we can proceed with the current approach given that it won't require any changes to existing OP components, and can refactor this into an enum and integrate everywhere later

File: crates/optimism/primitives/src/receipt.rs [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+            tx_type,
+            deposit_nonce,
+            deposit_receipt_version,
+        })
+    }
+}
+
+/// TODO: Remove once <https://github.com/alloy-rs/op-alloy/pull/317> is released.
+#[cfg(feature = "serde")]
+mod tx_type_serde {
```
Comment (mattsse) [inline crates/optimism/primitives/src/receipt.rs:250]: released now

File: crates/storage/codecs/derive/src/compact/mod.rs [hunk 3]
```diff
@@ -196,7 +196,7 @@ fn should_use_alt_impl(ftype: &str, segment: &syn::PathSegment) -> bool {
 pub fn get_bit_size(ftype: &str) -> u8 {
     match ftype {
         "TransactionKind" | "TxKind" | "bool" | "Option" | "Signature" => 1,
-        "TxType" => 2,
+        "TxType" | "OpTxType" => 2,
```
Comment (mattsse) [inline crates/storage/codecs/derive/src/compact/mod.rs:199]: this could be a footgun.

perhaps a better approach long term would be a trait, maybe.

**Review by mattsse — APPROVED** [review]
lgtm pending op-alloy bump

File: crates/storage/codecs/derive/src/compact/mod.rs (hunk 3 above)
Comment (klkvr) [reply crates/storage/codecs/derive/src/compact/mod.rs:199]: I've looked into this, we can't solve this with a trait while we depend on `modular_bitifield` the `get_bit_size` output is translates into `B2`/`B1` etc inputs for bitfield macro:
```rust
#[bitfield]
pub struct OpReceiptFlags {
    pub tx_type_len: B2,
    pub success_len: B1,
    pub cumulative_gas_used_len: B4,
    pub deposit_nonce_len: B1,
    pub deposit_receipt_version_len: B1,
    pub __zstd: B1,
    #[skip]
    unused: B6,
}
```

**Pull request merged**